File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersions.java
package co.infinum.princeofversions;

import android.content.Context;
import androidx.annotation.VisibleForTesting;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executor;

import javax.annotation.Nullable;

/**
 * This class represents main entry point for using library.
 * <p>
 * Most common way to create instance of this class should be using {@link Builder} or constructor with {@link Context} argument.
 * </p>
 * <p>
 * To check if update exists you can use two different approaches: synchronous and asynchronous.
 * </p>
 * <p>
 * Depending on used approach there are several versions of checkForUpdates method.
 * Synchronous execution is possible using checkForUpdates methods with one parameter ({@link Loader} or {@link String} as URL from which
 * update configuration will be downloaded.
 * Asynchronous execution expects one more parameter: {@link UpdaterCallback} callback through which results will be notified. Also,
 * there you can specify custom {@link Executor}, class which should run process, usually on background thread.
 * With this approach checkForUpdates method returns {@link PrinceOfVersionsCancelable} object which you can use to cancel request.
 * </p>
 * <p>
 * Here is code for most common usage of this library
 * <pre>
 *         {@link PrinceOfVersions} updater = new {@link PrinceOfVersions}(context);
 *         {@link PrinceOfVersionsCancelable} call = updater.checkForUpdates("http://example.com/some/update.json", callback); // starts
 *         checking
 * for update
 * </pre>
 */
public final class PrinceOfVersions {

    private final Presenter presenter;
    private final ApplicationConfiguration appConfig;
    private final Executor callbackExecutor;

    /**
     * Creates {@link PrinceOfVersions} using provided {@link Context}.
     *
     * @param context context which will be used for checking application version.
     */
    public PrinceOfVersions(Context context) {
        this(createDefaultParser(), createDefaultStorage(context),
            createDefaultCallbackExecutor(),
            createAppConfig(context));
    }

    @VisibleForTesting
    PrinceOfVersions(Storage storage, Executor callbackExecutor, ApplicationConfiguration appConfig) {
        this(createDefaultParser(), storage, callbackExecutor, appConfig);
    }

    @VisibleForTesting
    PrinceOfVersions(Storage storage, Executor callbackExecutor, ApplicationConfiguration appConfig,
        Map<String, RequirementChecker> checkers) {
        this(createMockedParser(checkers), storage, callbackExecutor, appConfig);
    }

    private PrinceOfVersions(ConfigurationParser configurationParser, Storage storage,
        Executor callbackExecutor, ApplicationConfiguration appConfig) {
        this.presenter = new PresenterImpl(
            new InteractorImpl(configurationParser),
            storage
        );
        this.callbackExecutor = callbackExecutor;
        this.appConfig = appConfig;
    }

    private static ConfigurationParser createDefaultParser(Map<String, RequirementChecker> requirementCheckers) {
        requirementCheckers.put(PrinceOfVersionsDefaultRequirementsChecker.KEY, new PrinceOfVersionsDefaultRequirementsChecker());
        return new JsonConfigurationParser(new PrinceOfVersionsRequirementsProcessor(
            requirementCheckers
        ));
    }

    private static ConfigurationParser createDefaultParser() {
        return new JsonConfigurationParser(new PrinceOfVersionsRequirementsProcessor());
    }

    private static ConfigurationParser createMockedParser(Map<String, RequirementChecker> requirementCheckers) {
        return new JsonConfigurationParser(new PrinceOfVersionsRequirementsProcessor(
            requirementCheckers
        ));
    }

    private static Storage createDefaultStorage(Context context) {
        return new PrinceOfVersionsDefaultNamedPreferenceStorage(context);
    }

    private static ApplicationConfiguration createAppConfig(Context context) {
        return new ApplicationConfigurationImpl(context);
    }

    private static Executor createDefaultCallbackExecutor() {
        return new PrinceOfVersionsCallbackExecutor();
    }

    /**
     * Start asynchronous check for update using provided {@link Executor} and {@link String}. Notifies result to provided {@link
     * UpdaterCallback}.
     *
     * @param url      Url from where update config will be loaded.
     * @param callback Callback to notify result.
     * @return instance through which is possible to cancel the call.
     */
    public PrinceOfVersionsCancelable checkForUpdates(String url, UpdaterCallback callback) {
        return checkForUpdates(new PrinceOfVersionsDefaultExecutor(), new NetworkLoader(url), callback);
    }

    /**
     * Start asynchronous check for update using provided {@link Executor} and {@link Loader}. Notifies result to provided {@link
     * UpdaterCallback}.
     *
     * @param loader   Instance for loading update config resource.
     * @param callback Callback to notify result.
     * @return instance through which is possible to cancel the call.
     */
    public PrinceOfVersionsCancelable checkForUpdates(Loader loader, UpdaterCallback callback) {
        return checkForUpdates(new PrinceOfVersionsDefaultExecutor(), loader, callback);
    }

    /**
     * Start asynchronous check for update using provided {@link Executor} and {@link String}. Notifies result to provided {@link
     * UpdaterCallback}.
     *
     * @param executor Instance for running check call.
     * @param url      Url from where update config will be loaded.
     * @param callback Callback to notify result.
     * @return instance through which is possible to cancel the call.
     */
    public PrinceOfVersionsCancelable checkForUpdates(Executor executor, String url, UpdaterCallback callback) {
        return checkForUpdates(executor, new NetworkLoader(url), callback);
    }

    /**
     * Start asynchronous check for update using provided {@link Executor} and {@link Loader}. Notifies result to provided {@link
     * UpdaterCallback}.
     *
     * @param executor Instance for running check call.
     * @param loader   Instance for loading update config resource.
     * @param callback Callback to notify result.
     * @return instance through which is possible to cancel the call.
     */
    public PrinceOfVersionsCancelable checkForUpdates(Executor executor, Loader loader, UpdaterCallback callback) {
        return checkForUpdatesInternal(executor, loader, new ExecutorUpdaterCallback(callback, callbackExecutor));
    }

    @VisibleForTesting
    PrinceOfVersionsCancelable checkForUpdatesInternal(Executor executor, Loader loader, UpdaterCallback callback) {
        return presenter.check(loader, executor, callback, appConfig);
    }

    /**
     * Start synchronous check for update using provided URL as {@link String}.
     *
     * @param url Url from where update config will be loaded.
     * @return result of update check.
     * @throws Throwable if error occurred.
     */
    public UpdateResult checkForUpdates(String url) throws Throwable {
        return checkForUpdates(new NetworkLoader(url));
    }

    /**
     * Start synchronous check for update using provided URL as {@link String}.
     *
     * @param loader Instance for loading update config resource.
     * @return result of update check.
     * @throws Throwable if error occurred.
     */
    public UpdateResult checkForUpdates(Loader loader) throws Throwable {
        return presenter.check(loader, appConfig);
    }

    /**
     * Creates new call object which will load configuration from specified url.
     *
     * @param url Url from where update config will be loaded.
     * @return Call with ability to execute or enqueue the check.
     */
    public PrinceOfVersionsCall newCall(String url) {
        return newCall(new NetworkLoader(url));
    }

    /**
     * Creates new call object which will load configuration from specified url.
     *
     * @param loader Instance for loading update config resource.
     * @return Call with ability to execute or enqueue the check.
     */
    public PrinceOfVersionsCall newCall(Loader loader) {
        return new UpdaterCall(this, loader);
    }

    /**
     * Helper class for building {@link PrinceOfVersions} object.
     * All methods are optional.
     */
    public static class Builder {

        private final Map<String, RequirementChecker> requirementCheckers = new HashMap<>();

        @Nullable
        private ConfigurationParser configurationParser;
        @Nullable
        private Storage storage;
        @Nullable
        private ApplicationConfiguration appConfig;
        @Nullable
        private Executor callbackExecutor;

        /**
         * Set a new configuration parser used to parse configuration file into the model.
         *
         * @param configurationParser Configuration parser
         * @return this builder
         */
        public Builder withParser(ConfigurationParser configurationParser) {
            this.configurationParser = configurationParser;
            return this;
        }

        /**
         * Set a new implementation of the storage used to store internal metadata about update check.
         *
         * @param storage Storage implementation
         * @return this builder
         */
        public Builder withStorage(Storage storage) {
            this.storage = storage;
            return this;
        }

        /**
         * Set a new callback executor which runs callback code on specific thread.
         *
         * @param callbackExecutor Callback executor
         * @return this builder
         */
        public Builder withCallbackExecutor(@Nullable final Executor callbackExecutor) {
            this.callbackExecutor = callbackExecutor;
            return this;
        }

        /**
         * Add a new custom requirements checker that's used in process of parsing JSON
         *
         * @param requirementsChecker Requirements checker
         * @param key                 Key with which we are going to get a wanted requirement from JSON
         * @return this builder
         */
        public Builder addRequirementsChecker(String key, RequirementChecker requirementsChecker) {
            this.requirementCheckers.put(key, requirementsChecker);
            return this;
        }

        /**
         * Remove custom requirements checker that's used in process of parsing JSON
         *
         * @param key                 Key with which we are going to get a wanted requirement from JSON
         * @return this builder
         */
        public Builder removeRequirementsChecker(String key) {
            this.requirementCheckers.remove(key);
            return this;
        }

        /**
         * Set a new application configuration, used to specify data about current application version
         *
         * @param appConfig Currect application configuration
         * @return this builder
         */
        @VisibleForTesting
        Builder withAppConfig(ApplicationConfiguration appConfig) {
            this.appConfig = appConfig;
            return this;
        }

        /**
         * Create the {@link PrinceOfVersions} instance using the configured values.
         *
         * @param context Context used to extract data about current application version
         * @return PrinceOfVersions instance
         */
        public PrinceOfVersions build(Context context) {
            return new PrinceOfVersions(
                configurationParser != null ? configurationParser : createDefaultParser(requirementCheckers),
                storage != null ? storage : createDefaultStorage(context),
                callbackExecutor != null ? callbackExecutor : createDefaultCallbackExecutor(),
                appConfig != null ? appConfig : createAppConfig(context)
            );
        }

        /**
         * Create the {@link PrinceOfVersions} instance using the configured values.
         * Note: both storage and application configuration has to be set to use this method.
         *
         * @return PrinceOfVersions instance
         */
        @VisibleForTesting
        PrinceOfVersions build() {
            if (storage == null || appConfig == null) {
                throw new UnsupportedOperationException(
                    "You must define storage and application configuration if you don't provide Context.");
            }
            return new PrinceOfVersions(
                configurationParser != null ? configurationParser : createDefaultParser(requirementCheckers),
                storage,
                callbackExecutor != null ? callbackExecutor : createDefaultCallbackExecutor(),
                appConfig
            );
        }
    }
}

File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersions.kt
package co.infinum.princeofversions

import android.content.Context
import androidx.annotation.VisibleForTesting
import java.util.concurrent.Executor

/**
 * This class represents main entry point for using library.
 *
 * Most common way to create instance of this class should be using [Builder] or constructor with [Context] argument.
 *
 * To check if update exists you can use two different approaches: synchronous and asynchronous.
 *
 * Depending on used approach there are several versions of checkForUpdates method.
 * Synchronous execution is possible using checkForUpdates methods with one parameter ([Loader] or [String] as URL from which
 * update configuration will be downloaded.
 * Asynchronous execution expects one more parameter: [UpdaterCallback] callback through which results will be notified. Also,
 * there you can specify custom [Executor], class which should run process, usually on background thread.
 * With this approach checkForUpdates method returns [PrinceOfVersionsCancelable] object which you can use to cancel request.
 *
 * Here is code for most common usage of this library
 * ```
 * PrinceOfVersions updater = new PrinceOfVersions(context);
 * PrinceOfVersionsCancelable call = updater.checkForUpdates("[http://example.com/some/update.json](http://example.com/some/update.json)", callback); // starts
 * checking
 * for update
 * ```
 */
class PrinceOfVersions private constructor(
    private val presenter: Presenter,
    private val appConfig: ApplicationConfiguration,
    private val callbackExecutor: Executor
) {

    /**
     * Creates a [PrinceOfVersions] instance using the provided [Context] and default dependencies.
     *
     * @param context The context used for checking the application version and for default storage.
     */
    constructor(context: Context) : this(
        presenter = PresenterImpl(
            InteractorImpl(createDefaultParser()),
            createDefaultStorage(context)
        ),
        appConfig = createAppConfig(context),
        callbackExecutor = createDefaultCallbackExecutor()
    )

    @VisibleForTesting
    internal constructor(storage: Storage, callbackExecutor: Executor, appConfig: ApplicationConfiguration) : this(
        presenter = PresenterImpl(
            InteractorImpl(createDefaultParser()),
            storage
        ),
        appConfig = appConfig,
        callbackExecutor = callbackExecutor
    )

    @VisibleForTesting
    internal constructor(
        storage: Storage,
        callbackExecutor: Executor,
        appConfig: ApplicationConfiguration,
        checkers: Map<String, RequirementChecker>
    ) : this(
        presenter = PresenterImpl(
            InteractorImpl(createMockedParser(checkers)),
            storage
        ),
        appConfig = appConfig,
        callbackExecutor = callbackExecutor
    )

    /**
     * Starts an asynchronous check for an update from a URL.
     *
     * @param url The URL from which the update configuration will be loaded.
     * @param callback The callback to notify with the result.
     * @return An instance through which the call can be canceled.
     */
    fun checkForUpdates(url: String, callback: UpdaterCallback): PrinceOfVersionsCancelable {
        return checkForUpdates(PrinceOfVersionsDefaultExecutor(), NetworkLoader(url), callback)
    }

    /**
     * Starts an asynchronous check for an update using a custom [Loader].
     *
     * @param loader The instance for loading the update configuration resource.
     * @param callback The callback to notify with the result.
     * @return An instance through which the call can be canceled.
     */
    fun checkForUpdates(loader: Loader, callback: UpdaterCallback): PrinceOfVersionsCancelable {
        return checkForUpdates(PrinceOfVersionsDefaultExecutor(), loader, callback)
    }

    /**
     * Starts an asynchronous check for an update from a URL on a custom [Executor].
     *
     * @param executor The instance for running the check call.
     * @param url The URL from which the update configuration will be loaded.
     * @param callback The callback to notify with the result.
     * @return An instance through which the call can be canceled.
     */
    fun checkForUpdates(executor: Executor, url: String, callback: UpdaterCallback): PrinceOfVersionsCancelable {
        return checkForUpdates(executor, NetworkLoader(url), callback)
    }

    /**
     * Starts an asynchronous check for an update using a custom [Loader] on a custom [Executor].
     *
     * @param executor The instance for running the check call.
     * @param loader The instance for loading the update configuration resource.
     * @param callback The callback to notify with the result.
     * @return An instance through which the call can be canceled.
     */
    fun checkForUpdates(executor: Executor, loader: Loader, callback: UpdaterCallback): PrinceOfVersionsCancelable {
        return checkForUpdatesInternal(executor, loader, ExecutorUpdaterCallback(callback, callbackExecutor))
    }

    @VisibleForTesting
    internal fun checkForUpdatesInternal(executor: Executor, loader: Loader, callback: UpdaterCallback): PrinceOfVersionsCancelable {
        return presenter.check(loader, executor, callback, appConfig)
    }

    /**
     * Starts a synchronous check for an update from a URL.
     *
     * @param url The URL from which the update configuration will be loaded.
     * @return The result of the update check.
     * @throws Throwable if an error occurs.
     */
    @Throws(Throwable::class)
    fun checkForUpdates(url: String): UpdateResult {
        return checkForUpdates(NetworkLoader(url))
    }

    /**
     * Starts a synchronous check for an update using a custom [Loader].
     *
     * @param loader The instance for loading the update configuration resource.
     * @return The result of the update check.
     * @throws Throwable if an error occurs.
     */
    @Throws(Throwable::class)
    fun checkForUpdates(loader: Loader): UpdateResult {
        return presenter.check(loader, appConfig)
    }

    /**
     * Creates a new call object that will load the configuration from the specified URL.
     *
     * @param url The URL from which the update configuration will be loaded.
     * @return A call with the ability to execute or enqueue the check.
     */
    fun newCall(url: String): PrinceOfVersionsCall {
        return newCall(NetworkLoader(url))
    }

    /**
     * Creates a new call object that will load the configuration using the specified [Loader].
     *
     * @param loader The instance for loading the update configuration resource.
     * @return A call with the ability to execute or enqueue the check.
     */
    fun newCall(loader: Loader): PrinceOfVersionsCall {
        return UpdaterCall(this, loader)
    }

    /**
     * A helper class for building a [PrinceOfVersions] object.
     * All methods are optional.
     */
    class Builder {

        private val requirementCheckers = mutableMapOf<String, RequirementChecker>()
        private var configurationParser: ConfigurationParser? = null
        private var storage: Storage? = null
        private var appConfig: ApplicationConfiguration? = null
        private var callbackExecutor: Executor? = null

        /**
         * Sets a custom configuration parser.
         */
        fun withParser(configurationParser: ConfigurationParser) = apply {
            this.configurationParser = configurationParser
        }

        /**
         * Sets a custom storage implementation.
         */
        fun withStorage(storage: Storage) = apply {
            this.storage = storage
        }

        /**
         * Sets a custom callback executor.
         */
        fun withCallbackExecutor(callbackExecutor: Executor?) = apply {
            this.callbackExecutor = callbackExecutor
        }

        /**
         * Adds a custom requirements checker.
         */
        fun addRequirementsChecker(key: String, requirementsChecker: RequirementChecker) = apply {
            this.requirementCheckers[key] = requirementsChecker
        }

        /**
         * Removes a custom requirements checker.
         */
        fun removeRequirementsChecker(key: String) = apply {
            this.requirementCheckers.remove(key)
        }

        @VisibleForTesting
        internal fun withAppConfig(appConfig: ApplicationConfiguration) = apply {
            this.appConfig = appConfig
        }

        /**
         * Creates the [PrinceOfVersions] instance using the configured values.
         */
        fun build(context: Context): PrinceOfVersions {
            val finalParser = this.configurationParser ?: createDefaultParser(requirementCheckers)
            val finalStorage = this.storage ?: createDefaultStorage(context)
            val finalCallbackExecutor = this.callbackExecutor ?: createDefaultCallbackExecutor()
            val finalAppConfig = this.appConfig ?: createAppConfig(context)

            return PrinceOfVersions(
                presenter = PresenterImpl(InteractorImpl(finalParser), finalStorage),
                appConfig = finalAppConfig,
                callbackExecutor = finalCallbackExecutor
            )
        }

        /**
         * Creates the [PrinceOfVersions] instance using the configured values.
         * Note: Both storage and application configuration must be set to use this method.
         */
        @VisibleForTesting
        internal fun build(): PrinceOfVersions {
            if (storage == null || appConfig == null) {
                throw UnsupportedOperationException(
                    "You must define storage and application configuration if you don't provide Context."
                )
            }
            val finalParser = this.configurationParser ?: createDefaultParser(requirementCheckers)
            val finalCallbackExecutor = this.callbackExecutor ?: createDefaultCallbackExecutor()

            return PrinceOfVersions(
                presenter = PresenterImpl(InteractorImpl(finalParser), storage!!),
                appConfig = appConfig!!,
                callbackExecutor = finalCallbackExecutor
            )
        }
    }

    private companion object {
        private fun createDefaultParser(requirementCheckers: MutableMap<String, RequirementChecker>): ConfigurationParser {
            requirementCheckers[PrinceOfVersionsDefaultRequirementsChecker.KEY] = PrinceOfVersionsDefaultRequirementsChecker()
            return JsonConfigurationParser(PrinceOfVersionsRequirementsProcessor(requirementCheckers))
        }

        private fun createDefaultParser(): ConfigurationParser {
            return createDefaultParser(mutableMapOf())
        }

        private fun createMockedParser(requirementCheckers: Map<String, RequirementChecker>): ConfigurationParser {
            return JsonConfigurationParser(PrinceOfVersionsRequirementsProcessor(requirementCheckers))
        }

        private fun createDefaultStorage(context: Context): Storage {
            return PrinceOfVersionsDefaultNamedPreferenceStorage(context)
        }

        private fun createAppConfig(context: Context): ApplicationConfiguration {
            return ApplicationConfigurationImpl(context)
        }

        private fun createDefaultCallbackExecutor(): Executor {
            return PrinceOfVersionsCallbackExecutor()
        }
    }
}
